waku/v2/api/publish/message_queue.go

```diff
@@ -0,0 +1,158 @@
+package publish
+
+import (
+	"container/heap"
+	"context"
+
+	"github.com/waku-org/go-waku/waku/v2/protocol"
+)
+
+// MessagePriority determines the ordering for the message priority queue
+type MessagePriority = int
+
+const (
+	LowPriority    MessagePriority = 1
+	MediumPriority MessagePriority = 2
+	HighPriority   MessagePriority = 3
+
+	DefaultPriority MessagePriority = MediumPriority
+)
+
+type envelopePriority struct {
+	envelope *protocol.Envelope
+	priority int
+	index    int
+}
+
+type envelopePriorityQueue []*envelopePriority
+
+func (pq envelopePriorityQueue) Len() int { return len(pq) }
+
+func (pq envelopePriorityQueue) Less(i, j int) bool {
+	if pq[i].priority > pq[j].priority {
+		return true
+	} else if pq[i].priority == pq[j].priority {
+		return pq[i].envelope.Message().GetTimestamp() < pq[j].envelope.Message().GetTimestamp()
+	}
+
+	return false
+}
+
+func (pq envelopePriorityQueue) Swap(i, j int) {
+	pq[i], pq[j] = pq[j], pq[i]
+	pq[i].index = i
+	pq[j].index = j
+}
+
+func (pq *envelopePriorityQueue) Push(x any) {
+	n := len(*pq)
+	item := x.(*envelopePriority)
+	item.index = n
+	*pq = append(*pq, item)
+}
+
+func (pq *envelopePriorityQueue) Pop() any {
+	old := *pq
+	n := len(old)
+	item := old[n-1]
+	old[n-1] = nil  // avoid memory leak
+	item.index = -1 // for safety
```

Comment:
I don't see the usage of index, should it be removed?

Reply:
It's used by `func (pq envelopePriorityQueue) Swap(i, j int)`

Reply:
ah, missed that one!